File: 2025/A/answers/java/Solution.java
import java.util.*;

public class Solution {
    static int[] parent;
    static boolean[][] usedEdgesMat;
    static boolean[] hasEdge;

    static int find(int i) {
        if (parent[i] == i) return i;
        return parent[i] = find(parent[i]);
    }

    static void unite(int i, int j) {
        int root_i = find(i);
        int root_j = find(j);
        if (root_i != root_j) parent[root_i] = root_j;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt()) return;
        int N = sc.nextInt();

        int P = sc.nextInt();
        int[] L = new int[P];
        for (int i = 0; i < P; i++) L[i] = sc.nextInt();

        int Q = sc.nextInt();
        int[] R = new int[Q];
        for (int i = 0; i < Q; i++) R[i] = sc.nextInt();

        if (N % 2 == 0) {
            if (N > 2) {
                System.out.println("N");
            } else {
                // N = 2, S = 2
                if (P + Q > 2) System.out.println("N");
                else if (P >= 2 && L[0] == L[1]) System.out.println("N");
                else if (Q >= 2 && R[0] == R[1]) System.out.println("N");
                else if (P >= 1 && Q >= 1 && L[0] == R[0]) System.out.println("N");
                else System.out.println("Y");
            }
            return;
        }

        long S = (long) N * (N - 1) / 2 + 1;
        if (P + Q > S) {
            System.out.println("N");
            return;
        }

        for (int i = 0; i < P - 1; i++) if (L[i] == L[i + 1]) { System.out.println("N"); return; }
        for (int i = 0; i < Q - 1; i++) if (R[i] == R[i + 1]) { System.out.println("N"); return; }

        if (P > 0 && Q > 0 && L[0] != R[Q - 1]) {
            System.out.println("N");
            return;
        }

        usedEdgesMat = new boolean[N + 1][N + 1];
        int usedCount = 0;
        for (int i = 0; i < P - 1; i++) {
            int u = L[i], v = L[i + 1];
            if (u > v) { int tmp = u; u = v; v = tmp; }
            if (usedEdgesMat[u][v]) { System.out.println("N"); return; }
            usedEdgesMat[u][v] = true;
            usedCount++;
        }
        for (int i = 0; i < Q - 1; i++) {
            int u = R[i], v = R[i + 1];
            if (u > v) { int tmp = u; u = v; v = tmp; }
            if (usedEdgesMat[u][v]) { System.out.println("N"); return; }
            usedEdgesMat[u][v] = true;
            usedCount++;
        }

        long E_prime = (S - 1) - usedCount;
        int u_target, v_target;
        if (P > 0 && Q > 0) { u_target = L[P - 1]; v_target = R[0]; }
        else if (P > 0) { u_target = L[P - 1]; v_target = L[0]; }
        else if (Q > 0) { u_target = R[Q - 1]; v_target = R[0]; }
        else { u_target = 1; v_target = 1; }

        if (E_prime == 0) {
            System.out.println(u_target == v_target ? "Y" : "N");
            return;
        }

        parent = new int[N + 1];
        hasEdge = new boolean[N + 1];
        for (int i = 1; i <= N; i++) parent[i] = i;
        for (int i = 1; i <= N; i++) {
            for (int j = i + 1; j <= N; j++) {
                if (!usedEdgesMat[i][j]) {
                    unite(i, j);
                    hasEdge[i] = hasEdge[j] = true;
                }
            }
        }

        int edgeComp = -1;
        for (int i = 1; i <= N; i++) {
            if (hasEdge[i]) {
                int c = find(i);
                if (edgeComp == -1) edgeComp = c;
                else if (edgeComp != c) { System.out.println("N"); return; }
            }
        }

        if (edgeComp != -1) {
            if (find(u_target) != edgeComp || find(v_target) != edgeComp) {
                System.out.println("N");
                return;
            }
        }

        System.out.println("Y");
    }
}

File: 2025/A/answers/kt/Solution.kt
import java.util.*

fun main() {
    val sc = Scanner(System.`in`)
    if (!sc.hasNextInt()) return
    val n = sc.nextInt()

    val p = sc.nextInt()
    val l = IntArray(p)
    for (i in 0 until p) l[i] = sc.nextInt()

    val q = sc.nextInt()
    val r = IntArray(q)
    for (i in 0 until q) r[i] = sc.nextInt()

    if (n % 2 == 0) {
        if (n > 2) {
            println("N")
        } else {
            // N = 2, S = 2
            if (p + q > 2) println("N")
            else if (p >= 2 && l[0] == l[1]) println("N")
            else if (q >= 2 && r[0] == r[1]) println("N")
            else if (p >= 1 && q >= 1 && l[0] == r[0]) println("N")
            else println("Y")
        }
        return
    }

    val s = n.toLong() * (n - 1) / 2 + 1
    if (p + q > s) {
        println("N")
        return
    }

    for (i in 0 until p - 1) if (l[i] == l[i + 1]) { println("N"); return }
    for (i in 0 until q - 1) if (r[i] == r[i + 1]) { println("N"); return }

    if (p > 0 && q > 0 && l[0] != r[q - 1]) {
        println("N")
        return
    }

    val usedEdgesMat = Array(n + 1) { BooleanArray(n + 1) }
    var usedCount = 0
    for (i in 0 until p - 1) {
        var u = l[i]
        var v = l[i + 1]
        if (u > v) { val tmp = u; u = v; v = tmp }
        if (usedEdgesMat[u][v]) { println("N"); return }
        usedEdgesMat[u][v] = true
        usedCount++
    }
    for (i in 0 until q - 1) {
        var u = r[i]
        var v = r[i + 1]
        if (u > v) { val tmp = u; u = v; v = tmp }
        if (usedEdgesMat[u][v]) { println("N"); return }
        usedEdgesMat[u][v] = true
        usedCount++
    }

    val ePrime = (s - 1) - usedCount
    val uTarget: Int
    val vTarget: Int
    if (p > 0 && q > 0) { uTarget = l[p - 1]; vTarget = r[0] }
    else if (p > 0) { uTarget = l[p - 1]; vTarget = l[0] }
    else if (q > 0) { uTarget = r[q - 1]; vTarget = r[0] }
    else { uTarget = 1; vTarget = 1 }

    if (ePrime == 0L) {
        println(if (uTarget == vTarget) "Y" else "N")
        return
    }

    val parent = IntArray(n + 1) { it }
    fun find(i: Int): Int {
        if (parent[i] == i) return i
        parent[i] = find(parent[i])
        return parent[i]
    }
    fun unite(i: Int, j: Int) {
        val rootI = find(i)
        val rootJ = find(j)
        if (rootI != rootJ) parent[rootI] = rootJ
    }

    val hasEdge = BooleanArray(n + 1)
    for (i in 1..n) {
        for (j in i + 1..n) {
            if (!usedEdgesMat[i][j]) {
                unite(i, j)
                hasEdge[i] = true
                hasEdge[j] = true
            }
        }
    }

    var edgeComp = -1
    for (i in 1..n) {
        if (hasEdge[i]) {
            val c = find(i)
            if (edgeComp == -1) edgeComp = c
            else if (edgeComp != c) { println("N"); return }
        }
    }

    if (edgeComp != -1) {
        if (find(uTarget) != edgeComp || find(vTarget) != edgeComp) {
            println("N")
            return
        }
    }

    println("Y")
}
